packages/in_app_purchase/ios/Classes/FIAPaymentQueueHandler.m

```diff
@@ -0,0 +1,128 @@
+
+@property(strong, nonatomic) SKPaymentQueue *queue;
+@property(nullable, copy, nonatomic) TransactionsUpdated transactionsUpdated;
+@property(nullable, copy, nonatomic) TransactionsRemoved transactionsRemoved;
+@property(nullable, copy, nonatomic) RestoreTransactionFailed restoreTransactionFailed;
+@property(nullable, copy, nonatomic)
+    RestoreCompletedTransactionsFinished paymentQueueRestoreCompletedTransactionsFinished;
+@property(nullable, copy, nonatomic) ShouldAddStorePayment shouldAddStorePayment;
+@property(nullable, copy, nonatomic) UpdatedDownloads updatedDownloads;
+
+@property(strong, nonatomic) NSMutableDictionary *transactionsSetter;
+
+@end
+
+@implementation FIAPaymentQueueHandler
+
+- (instancetype)initWithQueue:(nonnull SKPaymentQueue *)queue
+                     transactionsUpdated:(nullable TransactionsUpdated)transactionsUpdated
+                      transactionRemoved:(nullable TransactionsRemoved)transactionsRemoved
+                restoreTransactionFailed:(nullable RestoreTransactionFailed)restoreTransactionFailed
+    restoreCompletedTransactionsFinished:
+        (nullable RestoreCompletedTransactionsFinished)restoreCompletedTransactionsFinished
+                   shouldAddStorePayment:(nullable ShouldAddStorePayment)shouldAddStorePayment
+                        updatedDownloads:(nullable UpdatedDownloads)updatedDownloads {
+  self = [super init];
+  if (self) {
+    self.queue = queue;
+    self.transactionsUpdated = transactionsUpdated;
+    self.transactionsRemoved = transactionsRemoved;
+    self.restoreTransactionFailed = restoreTransactionFailed;
+    self.paymentQueueRestoreCompletedTransactionsFinished = restoreCompletedTransactionsFinished;
+    self.shouldAddStorePayment = shouldAddStorePayment;
+    self.updatedDownloads = updatedDownloads;
+    self.transactionsSetter = [NSMutableDictionary new];
+  }
+  return self;
+}
+
+- (void)addPayment:(SKPayment *)payment {
+  NSString *productID = payment.productIdentifier;
+  if (self.testing) {
+    productID = TestingProductID;
+  }
+  [self.queue addPayment:payment];
+}
+
+- (void)finishTransaction:(SKPaymentTransaction *)transaction {
+  @try {
+    [self.queue finishTransaction:transaction];
+  } @catch (NSException *e) {
+    // finish transaction will throw exception if the transaction type is purchasing. Raise the
+    // exception so the InAppPurchasePlugin will get notified.
+    [e raise];
+  }
+}
+
+#pragma mark - observing
+// Sent when the transaction array has changed (additions or state changes).  Client should check
+// state of transactions and finish as appropriate.
+- (void)paymentQueue:(SKPaymentQueue *)queue
+    updatedTransactions:(NSArray<SKPaymentTransaction *> *)transactions {
+  for (SKPaymentTransaction *transaction in transactions) {
+    [self.transactionsSetter setObject:transaction forKey:transaction.transactionIdentifier];
+  }
+  // notify dart through callbacks.
+  if (self.transactionsUpdated) {
```

Comment:
Should we even allow this to be constructed if we don't have a valid callback back up into dart for when our transactions are updated? Same general question for the other callbacks. I don't know if we really want this to be able to operate if it can't reach back up into the plugin with status updates.

Reply:
Maybe we can add all the call back when user adds the observer in dart. 
Or maybe we block the add payment flow if there is no dart observer been added? What do you think? Let's discuss this offline